### PR TITLE
Interpolation for parameters close to IRF boundaries

### PR DESCRIPTION
**Pull request by sona-patel**
Addressing issue #61

### CODE REVIEW

File: pyV2DL3/script/v2dl3_for_Eventdisplay.py [hunk 1]
```diff
@@ -52,6 +52,19 @@
...
+    nargs=1,
+    default=False,
+    help="IRF is extrapolated when parameter is found to be outside IRF range",
+)
+@click.option(
+    "--fuzzy_boundary",
+    nargs=1,
+    type=click.FLOAT,
+    default=0.0,
+    help="",
```
Comment (GernotMaier) [inline pyV2DL3/script/v2dl3_for_Eventdisplay.py:66]: Missing help message for fuzzy boundary.

What are the units of this value?

File: pyV2DL3/script/v2dl3_for_Eventdisplay.py [hunk 2]
```diff
@@ -52,6 +52,19 @@
     type=click.Path(exists=True),
     help="Load condition to filter events form json or yaml file.",
 )
+@click.option(
+    "--force_extrapolation",
+    nargs=1,
+    default=False,
```
Comment (GernotMaier) [inline pyV2DL3/script/v2dl3_for_Eventdisplay.py:58]: Is there a need for `type=click.BOOL` ?

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 3]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
 
 def check_parameter_range(par, par_irf, par_name):
-    """Check that coordinates are in range of provided IRF"""
+    """Check that coordinates are in range of provided IRF and whether extrapolation is to be done
+       0. checks if command line parameter force_extrapolation is given. If given,
+          the extrapolation will happen when parameter is outside IRF range. If parameter is
+          within IRF range, it works as normal. Default is False.
+       1. Further checks for fuzzy boundary (parameter close to boundary value).
+          If fuzzy boundary is within a given tolerance then IRF is interpolated for
+          at bounary value. Default is 0.0 tolerance.
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:47]: typo in `boundary`

File: pyV2DL3/eventdisplay/IrfInterpolator.py [hunk 4]
```diff
@@ -68,7 +69,12 @@ def __load_irf(self):
                 )
         self.irf_data = irf_data
         self.irf_axes = irf_axes
-        self.interpolator = RegularGridInterpolator(self.irf_axes, self.irf_data)
+        clk = click.get_current_context()
+        force_extrapolation = clk.params["force_extrapolation"]
+        if not force_extrapolation:
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/IrfInterpolator.py:74]: If not ... else conditions are difficult to read, could you first to 'if force extrapolation and then the second part? The else loop is a 'double negation' at this point, always hard to parse.

File: pyV2DL3/eventdisplay/IrfInterpolator.py [hunk 5]
```diff
@@ -68,7 +69,12 @@ def __load_irf(self):
                 )
         self.irf_data = irf_data
         self.irf_axes = irf_axes
-        self.interpolator = RegularGridInterpolator(self.irf_axes, self.irf_data)
+        clk = click.get_current_context()
+        force_extrapolation = clk.params["force_extrapolation"]
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/IrfInterpolator.py:73]: Why do you need to define a variable called `foce_extrapolation`? Couldn't you simply do `if clk.params["force_extrapolation"]` ?

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 6]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
 
 
 def check_parameter_range(par, par_irf, par_name):
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:40]: can we find a better name for the variable `par_irf`? It is very hard to see what it actually describe.

Or add a description of the arguments to the docstring? I have a hard time to see what is what.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 7]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
+        "\t{0} range of a given IRF: {1:.2f} - {2:.2f}".format(
             par_name, np.min(par_irf), np.max(par_irf)
         )
     )
+    clk = click.get_current_context()
+    force_extrapolation = clk.params["force_extrapolation"]
+    fuzzy_boundary = clk.params["fuzzy_boundary"]
     if np.all(par_irf < par) or np.all(par_irf > par):
-        raise ValueError("Coordinate not inside IRF {0} range".format(par_name))
+        if not force_extrapolation:
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:59]: Same comments as above - the `if not` makes the code hard to read.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 8]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
+            if fuzzy_boundary > 0.0:
+                fuzzy_diff_u = 1 - np.max(par_irf) / par
+                if fuzzy_diff_u < fuzzy_boundary:
+                    logging.warning(
+                        "Coordinate difference is {0:0.3f} of upper boundary value and is within {1:0.3f}".format(
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:64]: Very long line; please add a line break.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 9]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
+                            fuzzy_diff_u, fuzzy_boundary
+                        )
+                    )
+                    par = np.max(par_irf)
+                elif np.min(par_irf) != 0:
+                    fuzzy_diff_l = 1 - par / np.min(par_irf)
+                    if fuzzy_diff_l <= fuzzy_boundary:
+                        logging.warning(
+                            "Coordinate difference is {0:0.3f} of lower boundary value and is within {1:0.3f}".format(
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:73]: Very long line; please add a line break.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 10]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
     logging.info(
-        "\t{0} range of a given IRF: {1:.1f} - {2:.1f}".format(
+        "\t{0} range of a given IRF: {1:.2f} - {2:.2f}".format(
             par_name, np.min(par_irf), np.max(par_irf)
         )
     )
+    clk = click.get_current_context()
+    force_extrapolation = clk.params["force_extrapolation"]
+    fuzzy_boundary = clk.params["fuzzy_boundary"]
     if np.all(par_irf < par) or np.all(par_irf > par):
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:58]: Could we change this part to have not 4 blocks intended which makes it hard to read? Maybe easier to put some of the code in functions with very descriptive names?

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 11]
```diff
@@ -37,15 +38,50 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
-        raise ValueError("Coordinate not inside IRF {0} range".format(par_name))
+        if not force_extrapolation:
+            if fuzzy_boundary > 0.0:
+                fuzzy_diff_u = 1 - np.max(par_irf) / par
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:61]: Possible division by zero?

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Nice to see that this is coming along. Could you also add a unit test for `check_parameter_range` - especially for the case of the fuzzy edge interpolation? This would be very important to clarify the behavior in different cases.

File: pyV2DL3/eventdisplay/fillRESPONSE.py (hunk 10 above)
Comment (sona-patel) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:58]: Thank you for the reviewing this PR. I am not sure whether I understand this comment. Can you please elaborate a bit what changes are you asking to implement?

Comment (GernotMaier) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:58]: It is hard to read as there are too many levels of indentation.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 12]
```diff
@@ -37,53 +37,77 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
-        else:
-            logging.warning("IRF extrapolation allowed for coordinate not inside IRF {0} range".format(par_name))
+                raise ValueError(
+                    "Coordinate not inside IRF {0} range! Try using --fuzzy_boundary".format(par_name)
+                )
     return par
 
 
+def check_fuzzy_boundary(par, boundary, tolerance):
+    """" Checks if the paramters values is within the given tolerance.
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:78]: Typo: either singular `value` or `are` instead of `is`

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 13]
```diff
@@ -37,53 +37,77 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
+    boundary: lower or upper boundary value of stored IRF
+    tolerance: allowed value of --fuzzy_boundary command line argument
+
+    Returns
+    -------
+    Boolean value if tolerance is within given allowed value
+
+    """
+
+    fuzzy_diff = np.abs(boundary - par) / boundary
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:94]: is it possible that boundary is zero? In this case you have to check to prevent division by zero.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 14]
```diff
@@ -37,53 +37,77 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
-                                fuzzy_diff_l, fuzzy_boundary
-                            )
-                        )
-                        par = np.min(par_irf)
+    tolerance = clk.params["fuzzy_boundary"]
+    if np.all(irf_stored_par < par) or np.all(irf_stored_par > par):
+        if clk.params["force_extrapolation"]:
+            logging.warning("IRF extrapolation allowed for coordinate not inside IRF {0} range".format(par_name))
+        else:
+            if tolerance > 0.0:
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:61]: For readability, I suggest that you move this if-statement into the function check_fuzzy_boundary() (e.g., if tolerance <=0.0 return False) at the very beginning of the function.

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Much nicer now! 

Did you consider writing also a test for these functions?

File: pyV2DL3/eventdisplay/fillRESPONSE.py (hunk 13 above)
Comment (sona-patel) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:94]: boundary == 0, happens only for lower boundary, and that is taken care before calling this function for lower boundary value.

Comment (GernotMaier) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:94]: I would do that in check_fuzzy_boundary(), is this is actually the place where it matters. When reading the code, it is otherwise not clear why you check for `boundary != 0` (and you check for ` np.min(irf_stored_par) != 0`, which is only later assigned to boundary.

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 15]
```diff
@@ -36,16 +37,75 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
-    if np.all(par_irf < par) or np.all(par_irf > par):
-        raise ValueError("Coordinate not inside IRF {0} range".format(par_name))
+    clk = click.get_current_context()
+    tolerance = clk.params["fuzzy_boundary"]
+    if np.all(irf_stored_par < par) or np.all(irf_stored_par > par):
+        if clk.params["force_extrapolation"]:
+            logging.warning("IRF extrapolation allowed for coordinate not inside IRF {0} range".format(par_name))
+        else:
+            if tolerance > 0.0:
+                boundary = np.max(irf_stored_par)
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:62]: Suggest not the create a variable `boundary` here and in the else step, but put the `np.max(...)` directly in the function call of check_fuzzy_boundary. At this point, the reader sees twice the variable boundary, but it means different things! And you use the variable only once, so no need to define it.

**Review by GernotMaier — CHANGES_REQUESTED** [review]

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 16]
```diff
@@ -36,16 +37,75 @@ def print_logging_info(irf_to_store, camera_offsets, pedvar, zenith):
...
         )
     )
-    if np.all(par_irf < par) or np.all(par_irf > par):
-        raise ValueError("Coordinate not inside IRF {0} range".format(par_name))
+    clk = click.get_current_context()
+    tolerance = clk.params["fuzzy_boundary"]
+    if np.all(irf_stored_par < par) or np.all(irf_stored_par > par):
+        if clk.params["force_extrapolation"]:
+            logging.warning("IRF extrapolation allowed for coordinate not inside IRF {0} range".format(par_name))
+        else:
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:60]: Why do you need a block here like
```
else
   if
      ...
   else
     raise Value Error
```
why not simpler as
```
elif tolerance >0.
...
else
   raise ...
```

**Review by GernotMaier — CHANGES_REQUESTED** [review]

File: pyV2DL3/tests/test_fillRESPONSE.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import logging
+import pytest
+
+from pyV2DL3.eventdisplay.fillRESPONSE import check_fuzzy_boundary
+
+
+def test_check_fuzzy_boundary(caplog):
```
Comment (GernotMaier) [inline pyV2DL3/tests/test_fillRESPONSE.py:7]: Very nice! What about a test when par == boundary?

File: pyV2DL3/eventdisplay/fillRESPONSE.py [hunk 18]
```diff
@@ -91,21 +88,21 @@ def check_fuzzy_boundary(par, boundary, tolerance):
 
     """
 
-    fuzzy_diff = np.abs(boundary - par) / boundary
-    if fuzzy_diff < tolerance:
-        logging.warning(
-            "Coordinate tolerance is {0:0.3f} and is within {1:0.3f}".format(
-                fuzzy_diff, tolerance
+    if boundary != 0:
```
Comment (GernotMaier) [inline pyV2DL3/eventdisplay/fillRESPONSE.py:91]: Apologies to be so accurate here: minor improvement would be to have `if boundary == 0.: return False`, which would allow you to continue without intending a block.

Is there a default return value defined in this function?

**Review by GernotMaier — APPROVED** [review]
Very nice - I like it now, much more readable - thanks!

Two more minor comments which would be nice to have, otherwise we can merge it.

File: pyV2DL3/tests/test_fillRESPONSE.py (hunk 17 above)
Comment (sona-patel) [reply pyV2DL3/tests/test_fillRESPONSE.py:7]: `check_funzzy_boundary` function is not called  when `par == boundary` case occurs since this function is inside ` if np.all(irf_stored_par < par) or np.all(irf_stored_par > par):`.

Do you still suggest to have this case in test function?

File: pyV2DL3/eventdisplay/fillRESPONSE.py (hunk 18 above)
Comment (sona-patel) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:91]: No, the default return value is not defined for this function.  Yes, better to add `if boundary == 0.: return False`, so having this and defining default return value as `False`, would mean one of two mentioned conditions. This should be fine, right?

Comment (GernotMaier) [reply pyV2DL3/eventdisplay/fillRESPONSE.py:91]: Yes, this sounds right.

**Review by GernotMaier — APPROVED** [review]
All good, thank you!
